norddrop/src/ffi/uni.rs

```diff
@@ -0,0 +1,482 @@
+use std::{ffi::CStr, sync::Mutex, time::SystemTime};
+
+use drop_auth::SecretKey;
+use libc::{c_char, c_int, c_void};
+use slog::{o, Drain};
+
+use crate::{
+    device::{types, NordDropFFI},
+    ffi, Event,
+};
+
+pub type Result<T> = std::result::Result<T, crate::Error>;
+
+pub trait EventCallback: Send + Sync {
+    fn on_event(&self, event: Event);
+}
+
+pub trait Logger: Send + Sync {
+    fn on_log(&self, level: crate::LogLevel, msg: String);
+    fn level(&self) -> crate::LogLevel;
+}
+
+pub trait KeyStore: Send + Sync {
+    fn on_pubkey(&self, peer: String) -> Option<Vec<u8>>;
+    fn privkey(&self) -> Vec<u8>;
+}
+
+pub trait FdResolver: Send + Sync {
+    fn on_fd(&self, content_uri: String) -> Option<i32>;
+}
+
+pub struct NordDrop {
+    dev: Mutex<NordDropFFI>,
+    _cb: Box<CbContext>,
+    #[cfg(unix)]
+    fd_resolv: Mutex<Option<Box<FdResolverContext>>>,
```

Comment:
It must be `Mutex<Option<>>` because the `NordDrop` object must have only `fn(&self, ...)` methods. This is required by uniffy.
The double box is also needed because the C API required this `ctx: *mut c_void` to be passed into the callback. The thing is that Rust doesn't allow you to cast from `*mut c_void` to `mut dyn T`  because `Box<dyn ...>` is no longer a pointer but instead it is a [fat-pointer](https://stackoverflow.com/questions/57754901/what-is-a-fat-pointer). Hence the indirection needed to have `Box<T>` (without dyn) in the top-level structure.
Of course, it can be avoided with unsafe code, but since I'm going to remove the C API altogether I left it as it is since later on the problem would be nonexistent. Now it's just a limitation of the C API.